File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) {
            return null;
        }
        
        doLogic(root);
        
        return root;
    }
    
    private void doLogic(Node node) {
        Queue<Node> q = new LinkedList<>();
        q.add(node);
        
        while (!q.isEmpty()) {
            int n = q.size();
            
            Node prev = null;
            for (int i = 1; i <= n; i++) {
                Node rem = q.poll();
                
                if (i != 1) {
                    prev.next = rem;
                }
                
                if (rem.left != null) {
                    q.add(rem.left);
                }
                
                if (rem.right!=null) {
                    q.add(rem.right);
                }
                
                prev = rem;
            }
        }
    }
}

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	if root == nil {
		return nil
	}

	doLogic(root)

	return root
}

func doLogic(root *Node) {
	queue := NewQueue()
	queue.Push(root)

	for !queue.IsEmpty() {
		size := queue.Len()

		var prev *Node
		for i := 1; i <= size; i++ {
			current := queue.Pop()

			if i != 1 {
				prev.Next = current
			}

			if current.Left != nil {
				queue.Push(current.Left)
			}

			if current.Right != nil {
				queue.Push(current.Right)
			}

			prev = current
		}
	}
}

type Queue struct {
	v *list.List
}

func NewQueue() *Queue {
	return &Queue{list.New()}
}

func (q *Queue) Len() int {
	return q.v.Len()
}

func (q *Queue) IsEmpty() bool {
	if q.Len() == 0 {
		return true
	}

	return false
}

func (q *Queue) Push(v *Node) {
	q.v.PushBack(v)
}

func (q *Queue) Pop() *Node {
	front := q.v.Front()
	if front == nil {
		return nil
	}

	return q.v.Remove(front).(*Node)
}
